### PR TITLE
[EC-176] Implement ECIP-1017: monetary policy

### PR DESCRIPTION
**Pull request by rtkaczyk**
[ECIP-1017](https://github.com/ethereumproject/ECIPs/blob/master/ECIPs/ECIP-1017.md)

### CODE REVIEW

File: src/test/scala/io/iohk/ethereum/ledger/RewardTellerSpec.scala [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    val standardMP = MonetaryPolicyConfig(5000000, 0.2, 5000000000000000000L)
+
+    val testMP = MonetaryPolicyConfig(10, 0.5, 5000000)
+
+    val table = Table[MonetaryPolicyConfig, BigInt, List[BigInt], BigInt, List[BigInt]](
+      ("config", "blockNumber", "ommersNumbers", "expectedBlockReward", "expectedOmmersRewards"),
+      (standardMP, 1, Nil, 5000000000000000000L, Nil),
+      (standardMP, 1000000, List(999999), 5156250000000000000L, List(4375000000000000000L)),
+      (standardMP, 5000000, List(4999998, 4999997), 5312500000000000000L, List(3750000000000000000L, 3125000000000000000L)),
+      (standardMP, 5000001, Nil, 4000000000000000000L, Nil),
```
Comment (mcsherrylabs) [inline src/test/scala/io/iohk/ethereum/ledger/RewardTellerSpec.scala]: I just had  to add this, to see the changeover. -> `(standardMP, 5000000, Nil, 5000000000000000000L, Nil),`

File: src/test/scala/io/iohk/ethereum/ledger/RewardTellerSpec.scala [hunk 2]
```diff
@@ -0,0 +1,44 @@
+package io.iohk.ethereum.ledger
+
+import io.iohk.ethereum.utils.MonetaryPolicyConfig
+import org.scalatest.prop.PropertyChecks
+import org.scalatest.{FlatSpec, Matchers}
+
+class RewardTellerSpec extends FlatSpec with Matchers with PropertyChecks {
+
+  "RewardTeller" should "correctly calculate block and ommer rewards" in {
```
Comment (mcsherrylabs) [inline src/test/scala/io/iohk/ethereum/ledger/RewardTellerSpec.scala]: :+1:

File: src/test/scala/io/iohk/ethereum/network/handshaker/EtcHandshakerSpec.scala [hunk 3]
```diff
@@ -158,7 +158,7 @@ class EtcHandshakerSpec extends FlatSpec with Matchers  {
       override val customGenesisFileOpt: Option[String] = None
       override val daoForkBlockTotalDifficulty: BigInt = 0
       override val chainId: Byte = 0.toByte
-      override val blockReward: UInt256 = UInt256(0)
+      override val monetaryPolicyConfig: MonetaryPolicyConfig = null
```
Comment (mcsherrylabs) [inline src/test/scala/io/iohk/ethereum/network/handshaker/EtcHandshakerSpec.scala]: Is `???` more idiomatic instead of `null`?

File: src/main/scala/io/iohk/ethereum/ledger/RewardTeller.scala [hunk 4]
```diff
@@ -0,0 +1,69 @@
+package io.iohk.ethereum.ledger
+
+import io.iohk.ethereum.utils.MonetaryPolicyConfig
+
+/**
+  * Calculates rewards for mining blocks and ommers.
+  * Avoids floating point arithmetic. Because of that the formulas may look a bit unintuitive, but the important
+  * thing here is that we want to defer any division to be a single and final operation
+  */
+class RewardTeller(config: MonetaryPolicyConfig) {
```
Comment (mcsherrylabs) [inline src/main/scala/io/iohk/ethereum/ledger/RewardTeller.scala]: `RewardTeller` -> `BlockRewardCalculator`?

File: src/main/scala/io/iohk/ethereum/ledger/Ledger.scala [hunk 5]
```diff
@@ -24,6 +24,8 @@ trait Ledger {
 
 class LedgerImpl(vm: VM, blockchainConfig: BlockchainConfig) extends Ledger with Logger {
 
+  val rewardTeller = new RewardTeller(blockchainConfig.monetaryPolicyConfig)
```
Comment (mcsherrylabs) [inline src/main/scala/io/iohk/ethereum/ledger/Ledger.scala]: If `object EVMConfig` was `class EvmConfigure(blockChainConfig)` then LedgerImpl could be `LedgerImpl(vm, rewardTeller)`. `blockChainConfig` really gets around. But this is outside the scope of this PR I think.

File: src/main/scala/io/iohk/ethereum/ledger/RewardTeller.scala (hunk 4 above)
Comment (rtkaczyk) [inline src/main/scala/io/iohk/ethereum/ledger/RewardTeller.scala]: `BlockRewardCalculator` was my initial thought, but I like to give names that are a bit less obvious and more unique, especially for classes encapsulating important and isolated functionality. I find that it makes them stick in your head. I don't think I ever saw a `*Teller` before (think: bank teller), whereas `*Calculator` is so common. 

But if you don't buy it, I can change it.

File: src/main/scala/io/iohk/ethereum/ledger/RewardTeller.scala [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+      (ommerInclusionRewardDenom * eraDivisor)
+    baseReward + ommersReward
+  }
+
+  def calcOmmerMinerReward(blockNumber: BigInt, ommerNumber: BigInt): BigInt = {
+    val era = eraNumber(blockNumber)
+
+    if (era == 0) {
+      val numer = firstEraOmmerMiningRewardMaxNumer - (blockNumber - ommerNumber - 1)
+      firstEraBlockReward * numer / firstEraOmmerMiningRewardDenom
```
Comment (rtkaczyk) [inline src/main/scala/io/iohk/ethereum/ledger/RewardTeller.scala]: Oh, I hate those people... but OK 😄

File: src/test/scala/io/iohk/ethereum/ledger/RewardTellerSpec.scala (hunk 1 above)
Comment (rtkaczyk) [inline src/test/scala/io/iohk/ethereum/ledger/RewardTellerSpec.scala]: I will add this 👍

File: src/test/scala/io/iohk/ethereum/network/handshaker/EtcHandshakerSpec.scala (hunk 3 above)
Comment (rtkaczyk) [inline src/test/scala/io/iohk/ethereum/network/handshaker/EtcHandshakerSpec.scala]: `???` would only work with a `def` or a `lazy val`. In my view, `nulls` are fine in tests.

File: src/main/scala/io/iohk/ethereum/ledger/Ledger.scala (hunk 5 above)
Comment (rtkaczyk) [inline src/main/scala/io/iohk/ethereum/ledger/Ledger.scala]: I see your point, and yeah I wouldn't change it at this time. Currently the dependency between `Ledger` and `RewardTeller` feels like natural parent-child one, also reflected in the config.

File: src/main/scala/io/iohk/ethereum/ledger/RewardTeller.scala (hunk 4 above)
Comment (mcsherrylabs) [inline src/main/scala/io/iohk/ethereum/ledger/RewardTeller.scala]: As someone familiar with it, `RewardTeller` distinguishes it, but as a first time reader it clicked in my mind when I mentally renamed it `BlockRewardCalculator`, I'd buy `RewardCalculator` either as it has context where it's created. I have some hope that `BlockRewardCalculator` might become a ledger package level trait. (Also, the first word of the description is `Calculates...`)

File: src/test/scala/io/iohk/ethereum/network/handshaker/EtcHandshakerSpec.scala (hunk 3 above)
Comment (mcsherrylabs) [inline src/test/scala/io/iohk/ethereum/network/handshaker/EtcHandshakerSpec.scala]: :+1:

File: src/main/scala/io/iohk/ethereum/ledger/BlockRewardCalculator.scala [hunk 7]
```diff
@@ -0,0 +1,69 @@
+package io.iohk.ethereum.ledger
+
+import io.iohk.ethereum.utils.MonetaryPolicyConfig
+
+/**
+  * Calculates rewards for mining blocks and ommers.
+  * Avoids floating point arithmetic. Because of that the formulas may look a bit unintuitive, but the important
+  * thing here is that we want to defer any division to be a single and final operation
+  */
```
Comment (AlanVerbner) [inline src/main/scala/io/iohk/ethereum/ledger/BlockRewardCalculator.scala]: :heart: this doc

File: src/main/scala/io/iohk/ethereum/utils/Config.scala [hunk 8]
```diff
@@ -209,7 +208,21 @@ object BlockchainConfig {
...
+case class MonetaryPolicyConfig(
+  eraDuration: Int,
+  rewardRedutionRate: Double,
+  firstEraBlockReward: BigInt
+)
+
+object MonetaryPolicyConfig {
+  def apply(mpConfig: TypesafeConfig): MonetaryPolicyConfig = MonetaryPolicyConfig(
+    mpConfig.getInt("era-duration"),
+    mpConfig.getDouble("reward-reduction-rate"),
```
Comment (AlanVerbner) [inline src/main/scala/io/iohk/ethereum/utils/Config.scala]: I know it might be a strange/stupid case but If `reward-reduction-rate` is bigger than 1, then the `calculator.calcBlockMinerReward` returns a negative number which makes no sense. Should it be set as something like `Max(1, mpConfig.getDouble("reward-reduction-rate"))` ?

File: src/main/scala/io/iohk/ethereum/ledger/BlockRewardCalculator.scala [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+    val ommersReward = (firstEraBlockReward * ommersCount * ommerInclusionRewardNumer * eraMultiplier) /
+      (ommerInclusionRewardDenom * eraDivisor)
+    baseReward + ommersReward
+  }
+
+  def calcOmmerMinerReward(blockNumber: BigInt, ommerNumber: BigInt): BigInt = {
+    val era = eraNumber(blockNumber)
+
+    if (era == 0) {
+      val numer = firstEraOmmerMiningRewardMaxNumer - (blockNumber - ommerNumber - 1)
```
Comment (AlanVerbner) [inline src/main/scala/io/iohk/ethereum/ledger/BlockRewardCalculator.scala]: typo here?

Comment (rtkaczyk) [inline src/main/scala/io/iohk/ethereum/ledger/BlockRewardCalculator.scala]: Where exactly? `numer` is short for _numerator_ if that's what you mean.

File: src/main/scala/io/iohk/ethereum/utils/Config.scala (hunk 8 above)
Comment (rtkaczyk) [inline src/main/scala/io/iohk/ethereum/utils/Config.scala]: I don't like overcompensating for bad user/config input, but we could add a comment in the config file that the value should be in range [0.0, 1.0]. Would that be OK?

File: src/main/scala/io/iohk/ethereum/ledger/BlockRewardCalculator.scala (hunk 9 above)
Comment (AlanVerbner) [inline src/main/scala/io/iohk/ethereum/ledger/BlockRewardCalculator.scala]: Lol....it's ok...my bad

File: src/main/scala/io/iohk/ethereum/utils/Config.scala (hunk 8 above)
Comment (AlanVerbner) [inline src/main/scala/io/iohk/ethereum/utils/Config.scala]: Yes!!

Comment (mcsherrylabs) [inline src/main/scala/io/iohk/ethereum/utils/Config.scala]: There should be an `assert` in `MonetaryPolicyConfig` constructor for cases like this. I'll add this to the 'tidy config' task.
